Initial-databases

AfricanAlien opened the pull request:
Adds initial database setup and models
Sets up database tables for users, game instances, and player stats
Includes basic relationships and foreign key constraints

AfricanAlien committed:
Created initial framework for running flask website

AfricanAlien committed:
Merge branch 'main' of https://github.com/LinshanXiao/AWD-Project into initial-databases

AfricanAlien committed:
Defined tables for Users, League game instances and league game players

AfricanAlien committed:
Established relationship between tables and created tables

AfricanAlien committed:
Created script to generate tables

LinshanXiao commented on app/models.py:
```diff
@@ -0,0 +1,35 @@
+db = SQLAlchemy()
+
+class User(db.Model):
+    username = db.Column(db.String(150), nullable=False, unique=True, primary_key=True)
```
what do you think if we add a user id and set it as primary key instead of using username. Because user might want to change their username, also user id can be used as foreign key for other tables.

AfricanAlien replied:
yeah thats true changing should be an option. I have changed it

LinshanXiao commented on app/models.py:
```diff
@@ -0,0 +1,35 @@
+    game_id = db.Column(db.Integer, primary_key=True)
+    date_played = db.Column(db.DateTime, nullable=False)
+    game_duration = db.Column(db.Time, nullable=False)
+    winning_team = db.Column(db.String(150), nullable=False)
```
I see you are trying to create relationships between users and other tables. How about writing something like: 

players = db.relationship('League_Game_Player', back_populates='game')

AfricanAlien replied:
I transitioned it all to the newer format that the lab uses so it looks a bit different but it is now segmented as you suggested instead of 1 relationship definition

LinshanXiao commented on app/models.py:
```diff
@@ -0,0 +1,35 @@
+    league_username = db.Column(db.String(150), nullable=True, unique=True)
+    valorant_username = db.Column(db.String(150), nullable=True, unique=True)
+    PUBG_username = db.Column(db.String(150), nullable=True, unique=True)
+    apex_username = db.Column(db.String(150), nullable=True, unique=True)
```
league_game_players = db.relationship('League_Game_Player', back_populates='user')

LinshanXiao commented on app/models.py:
```diff
@@ -0,0 +1,35 @@
+    deaths = db.Column(db.Integer, nullable=False)
+    assists = db.Column(db.Integer, nullable=False)
+    kda = db.Column(db.Float, nullable=False)
+    team = db.Column(db.String(150), nullable=False)
```
user = db.relationship('User', back_populates='league_game_players')
game = db.relationship('League_Game_Instance', back_populates='players')

LinshanXiao commented on app/models.py:
```diff
@@ -0,0 +1,35 @@
+    secondary=League_Game_Player.__table__,
+    backref='user',
+    lazy='dynamic'
+)
```
Just providing an idea, but you decide which is better for you to use and easier to implement. :)

AfricanAlien replied:
as you suggested having it split up does make it more readable so ive updated

LinshanXiao commented on app/routes.py:
```diff
@@ -1,7 +1,8 @@
 
-@application.route('/')
-@application.route('/index')
+
```
will need to add some routes later on, so our database can interact with the client.

LinshanXiao reviewed (commented):
Some of my thoughts, hope that make sense. 
And yes, please do change some variable names. Thank you ~

LinshanXiao commented on create_tables.py:
```diff
@@ -0,0 +1,5 @@
+from app.models import db
+
+with app.app_context():
+    db.create_all()
```
NOTE: refrain from using db.create_all() in the final project.
This method does not track changes to the models or generate migration scripts.
Using it may cause schema mismatches, and it bypasses Flask-Migrate,
which is required for this project (we must demonstrate at least two migrations).
Please use `flask db migrate` and `flask db upgrade` instead to ensure database versioning and data integrity.

LinshanXiao committed:
making login page work

LinshanXiao committed:
Delete settings.json

fix little bug due to accidently used liver server

LinshanXiao committed:
adding jinja templates

1. creating a base html and move main.html to home.html
2.  implement jinja templates so that user clicking each tab will stay in same page
3.before our database is complete, I created a fake login user to test.

AfricanAlien committed:
Created config.py to configure database path for SQLAlchemy

AfricanAlien committed:
Updated database models to comply with newer SQLAlchemy 2.0 syntax

AfricanAlien committed:
Fixed database date and time values not being a python type

AfricanAlien committed:
added .db to gitignore

AfricanAlien committed:
Created migrations for database

AfricanAlien committed:
Removed setup.py as run.py also exists

AfricanAlien committed:
Merge branch 'implement-jinja' of https://github.com/LinshanXiao/AWD-Project into initial-databases

AfricanAlien committed:
Updated variable names to comply

AfricanAlien committed:
Changed url for back button from index to login page

AfricanAlien committed:
Added database integration for register page and fixed syntax in models

LinshanXiao committed:
delete create_tables.py otherwise it will throw an error when user register an account, and make gitigonore more flexible for ignore database file, then, change routes.py to make login work as make it read user info through databse

LinshanXiao committed:
created a requirements.txt to store the packages or libraries we need for our application
